services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -0,0 +1,231 @@
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.protocols.core.OperationInfo;
+import software.amazon.awssdk.protocols.core.PathMarshaller;
+import software.amazon.awssdk.protocols.core.ProtocolUtils;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.internal.handlers.EndpointAddressInterceptor;
+import software.amazon.awssdk.services.s3.model.GetUrlRequest;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Utilities for working with Amazon S3 objects. An instance of this class can be created by:
+ * 1) Directly using the {@link #builder()} method. You have to manually specify the configuration params like region,
+ * s3Configuration on the builder.
+ * 2) Using the low-level client {@link S3Client#utilities()} method. This is recommended as SDK will use the same
+ * configuration from the {@link S3Client} object to create the {@link S3Utilities} object.
+ *
+ * Note: This class should not have methods that make network calls.
+ */
+@Immutable
+@SdkPublicApi
+public final class S3Utilities {
+
+    private final Region region;
+
+    private final S3Configuration s3Configuration;
+
+    /**
+     * SDK currently validates that region is present while constructing {@link S3Utilities} object.
+     * This can be relaxed in future when more methods are added that doesn't use region.
+     */
+    public S3Utilities(Builder builder) {
+        this.region = Validate.paramNotNull(builder.region, "Region");
+        this.s3Configuration = builder.s3Configuration;
+    }
+
+    /**
+     * Creates a builder for {@link S3Utilities}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    // Used by low-level client
+    @SdkInternalApi
+    static S3Utilities create(SdkClientConfiguration clientConfiguration) {
+        return S3Utilities.builder()
+                          .region(clientConfiguration.option(AwsClientOption.AWS_REGION))
+                          .s3Configuration((S3Configuration) clientConfiguration.option(SdkClientOption.SERVICE_CONFIGURATION))
+                          .build();
+    }
+
+    /**
+     * Returns the URL for an object stored in Amazon S3.
+     *
+     * If the object identified by the given bucket and key has public read permissions,
+     * then this URL can be directly accessed to retrieve the object's data.
+     *
+     * <p>
+     *     If same configuration options are set on both #GetUrlRequest and #S3Utilities objects (for example: region),
+     *     the configuration set on the #GetUrlRequest takes precedence.
+     * </p>
+     *
+     * @param getUrlRequest request to construct url
+     * @return A URL for an object stored in Amazon S3.
+     * @throws MalformedURLException Generated Url is malformed
+     */
+    public URL getUrl(GetUrlRequest getUrlRequest) throws MalformedURLException {
```

Comment:
Any reason not to have `GetUrlResponse` for symmetry and extensibility?

Reply:
I think there was some pushback in the design meeting (my initial pr had `GetUrlResponse`) as the only use case for this api was to generate a url.

Reply:
I like response objects, since changing return types is impossible, and adding a new method with a different return type will always be confusing, but I don't want to reopen old wounds.

Reply:
I don't see us adding a new method with just `GetUrlRequest` type as input.